nb2/bot/slack_bot.py

```diff
@@ -194,113 +241,177 @@ def remember(self, target_slack_user_id: str, quote: str):
+            nostalgia_user_target: string representing slack_user_id, ghost_user_id or display_name
+                                   for Person to retrieve quote.
+                                   If not provided, this will return a random quote from
+                                   a random Person.
 
         Returns:
-            A tuple with a person's name and a Quote (or None if no quotes were found)
+            A tuple. If the Person doesn't exist, the tuple will be a string with a string
+            identifier and None. Otherwise, the tuple will contain a Person and a Quote.
         """
-        person = get_person_name_by_slack_user_id(nostalgia_user_target)
-        user_info = self.fetch_user_info(nostalgia_user_target)
-        real_name = user_info.get("real_name")
+        if not nostalgia_user_target:
+            quote = get_random_quote_from_any_person()
+            person = get_person_by_quote(quote)
+            return person, quote
 
+        person, is_active = get_person(nostalgia_user_target)
+
+        # The Person doesn't exist in the database, so return their first name,
+        # or the provided nostalgia_user_target with None as the Quote.
         if person is None:
-            return real_name, None
+            return nostalgia_user_target, None
+        if is_active:
+            person = self.update_person(person)
 
-        quote = get_random_quotes_from_person(nostalgia_user_target)
+        quote = get_random_quotes_from_person(person)
 
+        # The Person doesn't have any quotes saves so return their first name,
+        # with None as the Quote.
         if not quote:
-            return real_name, None
+            return nostalgia_user_target, None
 
+        # Success!
         return person, quote[0]
 
-    def quote(self, nostalgia_user_target: str):
+    def quote(self, message: str) -> Result:
         """
-        Recall a quote from NB's memory of a Person with nostalgia_user_target, if they exist.
+        Recall a quote from NB's memory of a Person, if they exist.
 
         Args:
-            nostalgia_user_target: string representing slack user id for Person
-                                   to retrieve quote
+            message: A message that was sent to the bot.
 
         Returns:
             A Result namedtuple.
         """
-        real_name, quote = self._random_quote(nostalgia_user_target)
+        matched = re.match(self.QUOTE_REGEX, message, re.I)
+
+        nostalgia_user_target = trim_mention(matched.group("nostalgia_user_target"))
+
+        try:
+            person, quote = self._random_quote(nostalgia_user_target)
+        except MultipleResultsFound:
+            return self.Result(
+                ok=True,
+                message=(
+                    f"I know more than one person under the name {nostalgia_user_target}. ",
+                    'Please specify them by their username in the format of "initial + last name"',
```

Comment:
We should probably update this to either use "display_name", "full name" or ghost_user_id (which they could get from an admin. But I'm not sure we can even hit this error right now because names won't match if they contain spaces and ghost_user_ids are always unique

Reply:
What do you mean by full name? We don't support looking up by first + last name yet right?

I think we'd only ever get here if the user tries to use display name because @ uses slack IDs, which are unique, and only display names could have duplicates, so the only option left is `ghost_user_id`

Reply:
Yes but theoretically we could have two users with the same display names and one of them has a display_name == ghost_user_id. Since we do the display_name check before the ghost_user_id check we could end up with two users in this case.

Reply:
we could change `get_person` to return get_person_by_ghost_user_id(user_id) or get_person_by_display_name(user_id) so that we avoid that

Reply:
But in that case if we want to quote the person whose display name != ghost_user_id, we will never be able to get there because it'll always return the other person first and never get to the part where it gets people by display name right :thinking:

Reply:
I think eventually this would resolve itself when we do the fancy reaction/option buttons that allow the user to choose on the spot which person it is, and since this isn't something that's supported in nb1 currently we could leave it as is for now?

Reply:
We can probably just update the message to include the display name. Not that I think our pattern matching will handle display names with spaces.
The case I mentioned is probably super unlikely